source/common/upstream/health_checker_impl.cc

```diff
@@ -279,15 +282,50 @@ void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onInterval() {
 
 void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onResetStream(Http::StreamResetReason,
                                                                         absl::string_view) {
+  request_in_flight_ = false;
+  ENVOY_CONN_LOG(debug, "connection/stream error health_flags={}", *client_,
+                 HostUtility::healthFlagsToString(*host_));
+  if (client_ && received_no_error_goaway_) {
+    client_->close();
+  }
+
   if (expect_reset_) {
     return;
   }
 
-  ENVOY_CONN_LOG(debug, "connection/stream error health_flags={}", *client_,
-                 HostUtility::healthFlagsToString(*host_));
   handleFailure(envoy::data::core::v3::NETWORK);
 }
 
+void HttpHealthCheckerImpl::HttpActiveHealthCheckSession::onGoAway(
+    Http::GoAwayErrorCode error_code) {
+  ENVOY_CONN_LOG(debug, "connection going away health_flags={}", *client_,
+                 HostUtility::healthFlagsToString(*host_));
+
+  // If GOAWAY handling is explicitly disabled via runtime, go back to the old
+  // behavior of ignoring GOAWAY completely.
+  if (parent_.runtime_.snapshot().featureEnabled("health_check.disable_goaway_handling", 0UL)) {
```

Comment:
@alyssawilk it turned out to be simple to go back to the old behavior. Let me know if I got the defaults the right direction, not sure if you wanted to make this opt in or opt out

Reply:
opt-out is fine, but can you please check https://github.com/envoyproxy/envoy/blob/master/CONTRIBUTING.md#runtime-guarding
and go with the naming and such used there? 
If you stick the flag in runtime_features.cc it'll go back to opt-out.